osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -255,7 +257,76 @@ private EncryptedTypedSecret retrieveSecret(
     return Optional.of(secretsList.iterator().next());
   }
 
-  public synchronized void storeSecrets(
+  public synchronized void storeOrResetNewSecrets(
+      final String deviceIdentification, final List<TypedSecret> typedSecrets) {
+    final List<TypedSecret> typedNewSecretsToStore = new ArrayList<>();
+    final List<TypedSecret> typedNewSecretsToReset = new ArrayList<>();
+    for (final TypedSecret typedSecret : typedSecrets) {
+      final boolean thereAreNoKeysWithStatusNew =
+          this.thereAreNoKeysWithStatusNew(deviceIdentification, typedSecret.getSecretType());
+      if (thereAreNoKeysWithStatusNew) {
+        typedNewSecretsToStore.add(typedSecret);
+      } else {
+        typedNewSecretsToReset.add(typedSecret);
+      }
+    }
+    this.storeSecrets(deviceIdentification, typedNewSecretsToStore);
+    typedNewSecretsToReset.forEach(
+        ts -> this.resetNewSecret(deviceIdentification, ts.getSecretType()));
+  }
+
+  private boolean thereAreNoKeysWithStatusNew(
+      final String deviceIdentification, final SecretType secretType) {
+    final int numberOfNewSecrets =
+        this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW);
+    return numberOfNewSecrets == 0;
+  }
+
+  private TypedSecret resetNewSecret(
+      final String deviceIdentification, final SecretType secretType) {
+    // the most recent key of all NEW keys of the device and type gets a new creation date and is
+    // returned. All other NEW keys of the device and type are set to status EXPIRED.
+    DbEncryptedSecret mostRecentFoundSecret = null;
+    final List<DbEncryptedSecret> foundSecrets =
+        this.secretRepository.findSecrets(deviceIdentification, secretType, SecretStatus.NEW);
+
+    boolean firstInListAndMostRecent = true;
+    for (final DbEncryptedSecret foundSecret : foundSecrets) {
+      if (firstInListAndMostRecent) {
+        foundSecret.setCreationTime(new Date());
```

Comment:
This indeed was used to prevent simulataneous execution of multiple key changing processes. This is now solved in a different way not using the NEW key's creation date. Therefor resetting this date is obsolete and is removed